Archives/ClmFile.cpp

```diff
@@ -2,104 +2,69 @@
 #include "../XFile.h"
 #include <stdexcept>
 #include <algorithm>
-#include <array>
 
 namespace Archives
 {
 	const uint32_t CLM_WRITE_SIZE = 0x00020000;
-	const uint32_t RIFF = 0x46464952;	// "RIFF" Resource Interchange File Format
-	const uint32_t WAVE = 0x45564157;	// "WAVE"
-	const uint32_t FMT  = 0x20746D66;	// "fmt "
-	const uint32_t DATA = 0x61746164;	// "data"
 
 	ClmFile::ClmFile(const char *fileName) : ArchiveFile(fileName)
 	{
-		m_FileName = nullptr;
-		m_IndexEntry = nullptr;
-
-		// Open the file for reading
-		m_FileHandle = CreateFileA(fileName,					// filename
-			GENERIC_READ,				// desired access
-			FILE_SHARE_READ,			// share mode
-			nullptr,					// security attributes
-			OPEN_EXISTING,				// creation disposition
-			FILE_ATTRIBUTE_NORMAL,		// attributes
-			nullptr);					// template
-
-		if (m_FileHandle == INVALID_HANDLE_VALUE) {
+		try {
+			clmFileReader = std::make_unique<FileStreamReader>(fileName);
+		}
+		catch (std::exception& e) {
 			throw std::runtime_error("Could not open clm file " + std::string(fileName) + ".");
 		}
 
-		if (ReadHeader() == false) {
-			throw std::runtime_error("Invalid clm header in " + std::string(fileName) + ".");
+		try {
+			ReadHeader();
+		} 
+		catch (std::exception& e) {
+			throw std::runtime_error("Invalid clm header read from file " + std::string(fileName) + ". Internal error message: " + e.what() );
 		}
-
-		// Initialize the unknown data
-		memset(m_Unknown, 0, 6);
-		m_Unknown[4] = 1;
 	}
 
-	ClmFile::~ClmFile()
-	{
-		delete[] m_FileName;
-		delete[] m_IndexEntry;
-
-		if (m_FileHandle) {
-			CloseHandle(m_FileHandle);
-		}
-	}
+	ClmFile::~ClmFile() { }
 
 
 	// Reads in the header when the volume is first opened and does some
 	// basic error checking on the header.
-	// Returns nonzero if the header was read successfully and is intact
-	// Returns zero if an error occured
-	bool ClmFile::ReadHeader()
+	// Throws an error is problems encountered while reading the header.
+	void ClmFile::ReadHeader()
 	{
-		char buff[32];
-		unsigned long numBytes;
-		int i;
-
-		if (ReadFile(m_FileHandle, buff, 32, &numBytes, nullptr) == 0) return false;
-		if (memcmp(buff, "OP2 Clump File Version 1.0\x01A\0\0\0\0", 32)) return false;
+		clmFileReader->Read((char*)&clmHeader, sizeof(ClmHeader));
+		
+		if (!clmHeader.CheckFileVersion()) {
```

Comment:
I suppose we can have both a check and a verify function. The verify function can just wrap the check function and throw the exception.

Reply:
That allows the most flexibility. I often do that.

The decision for me, is based on possible use cases. If checking something results in a normal control flow decision, you should have a method returning a value. It's bad design to use exceptions for normal program flow control. If that happens, or is likely to happen, you need a function with a proper return value. On the other hand, if you just need something to be true to continue, otherwise abort with an error, an exception raising function should be provided. Some checks are a bit split where it depends on context as to whether it constitutes a normal control flow decision, or an unexpected error condition. Those are the cases where I like to provide both, with one implemented in terms of the other.

My gut feeling here is there isn't a whole lot of reason for checking that doesn't lead to aborting with an error. Perhaps a get method, where the current value can be displayed, but the only time I see a check happening is when we abort further processing.